<?xml version="1.0"?>
<robot name="kinect" xmlns:xacro="http://ros.org/wiki/xacro">
  <xacro:macro name="kinect_xacro" params="name parent *origin">
    <property name="M_PI" value="3.14159"/>
    <joint name="${name}_joint" type="fixed">
      <xacro:insert_block name="origin"/>
      <parent link="${parent}"/>
      <child link="${name}_link" />
    </joint>
    
    <link name="${name}_link">
      <visual>
         <!-- NOTE: I (Georg Bartels) put these to be visually pleasing with our calibration
                    on February 11th, 2015. They are not set in stone! If the calibration
                    changes you might have to touch them again. Sorry! -->
        <origin xyz="-0.05 -0.17 -0.02" rpy="${M_PI/2.0} 0 ${M_PI/2.0}"/>
        <geometry>
          <mesh filename="package://gait_training_robot/meshes/68-04546_Kinect_Sensor--coarse.STL" scale="0.001 0.001 0.001" />
        </geometry>
        <material name="gray">
          <color rgba="0.5 0.5 0.5 1"/>
        </material>
      </visual>
    </link>
      <!-- NOTE: I (Zhuo Chen) added some dark cyclinders to highlight the orientation the Kinect! -->
    <material name="lens_color">
      <color rgba="0.1 0.1 0.1 1"/>
    </material>

    <!-- lens1 -->
    <link name="${name}_lens1_link">
      <visual>
        <origin xyz="0. 0. -0.02" rpy="${M_PI/2.0} 0 ${M_PI/2.0}"/>
        <geometry>
          <cylinder radius="0.01" length="0.005" />
        </geometry>
        <material name="lens_color">
          <color rgba="0.5 0.5 0.5 1"/>
        </material>
      </visual>
    </link>
    <joint name="${name}_lens1_joint" type="fixed">
      <origin xyz="0.010 -0.02 0.02" rpy="0.0 0.0 0.0"/>
      <parent link="${name}_link"/>
      <child link="${name}_lens1_link"/>
    </joint>

    <!-- lens2 -->
    <link name="${name}_lens2_link">
      <visual>
        <origin xyz="0. 0. -0.02" rpy="${M_PI/2.0} 0 ${M_PI/2.0}"/>
        <geometry>
          <cylinder radius="0.01" length="0.005" />
        </geometry>
        <material name="lens_color">
          <color rgba="0.5 0.5 0.5 1"/>
        </material>
      </visual>
    </link>
    <joint name="${name}_lens2_joint" type="fixed">
      <origin xyz="0.010 -0.045 0.02" rpy="0.0 0.0 0.0"/>
      <parent link="${name}_link"/>
      <child link="${name}_lens2_link"/>
    </joint>

    <!-- lens3 -->
    <link name="${name}_lens3_link">
      <visual>
        <origin xyz="0. 0. -0.02" rpy="${M_PI/2.0} 0 ${M_PI/2.0}"/>
        <geometry>
          <cylinder radius="0.01" length="0.005" />
        </geometry>
        <material name="lens_color">
          <color rgba="0.5 0.5 0.5 1"/>
        </material>
      </visual>
    </link>
    <joint name="${name}_lens3_joint" type="fixed">
      <origin xyz="0.010 -0.095 0.02" rpy="0.0 0.0 0.0"/>
      <parent link="${name}_link"/>
      <child link="${name}_lens3_link"/>
    </joint>
    
    <!--<joint name="${name}_rgb_joint" type="fixed">
      <origin xyz="0.0 -0.045 0.0" rpy="0.0 0.0 0.0"/>
      <parent link="${name}_link"/>
      <child link="${name}_rgb_frame"/>
    </joint>
    
    <link name="${name}_rgb_frame"/>
    
    <joint name="${name}_rgb_optical_joint" type="fixed">
      <origin xyz="0.0 0.0 0.0" rpy="${-M_PI/2.0} 0.0 ${-M_PI/2.0}"/>
      <parent link="${name}_rgb_frame"/>
      <child link="${name}_rgb_optical_frame"/>
    </joint>
    
    <link name="${name}_rgb_optical_frame"/>
    
    <joint name="${name}_depth_joint" type="fixed">
      <origin xyz="0.0 -0.02 0.0" rpy="0.0 0.0 0.0"/>
      <parent link="${name}_link"/>
      <child link="${name}_depth_frame"/>
    </joint>
    
    <link name="${name}_depth_frame"/>
    
    <joint name="${name}_depth_optical_joint" type="fixed">
      <origin xyz="0.0 0.0 0.0" rpy="${-M_PI/2.0} 0.0 ${-M_PI/2.0}"/>
      <parent link="${name}_depth_frame"/>
      <child link="${name}_depth_optical_frame"/>
    </joint>
    
    <link name="${name}_depth_optical_frame"/>-->

  </xacro:macro>
</robot>
